healthbox3/main.py

```diff
@@ -9,172 +20,241 @@
+        return json.dumps(self.__config)
 
     @om_expose
     def set_config(self, config):
         config = json.loads(config)
-        for key in config:
-            if isinstance(config[key], six.string_types):
-                config[key] = str(config[key])
-        self._config_checker.check_config(config)
-        self._config = config
-        self._read_config()
+        try:
+            self._check_config(config)
+        except Exception as ex:
+            self.logger("Could not set new config, config check failed: {}".format(ex))
+            return json.dumps({'success': False})
+
+        self.__config_checker.check_config(config)
         self.write_config(config)
+        self.__config = config
+        self.logger("Succesfully saved new config")
+
         return json.dumps({'success': True})
+
+    @staticmethod
+    def _check_config(config):
+        # check if all fields are populated
+        if 'update_delay' not in config:
+            raise RuntimeError('Config has no field "update_delay" (required)')
+
+        # check all the fields if they are valid input
+        delay = config['update_delay']
+        try:
+            delay = int(delay)
+        except Exception as ex:
+            raise RuntimeError('field "update_delay": ({}) is not a valid input in the provided config (not int)'.format(delay))
+
+    @background_task
+    def background_worker(self):
+        while True:
+            self.api_handler.do_requests()
+            time.sleep(10)
+
+    def discover_callback(self, ip):
+        # type: (str) -> None
+        """ callback for when a new device has been discovered """
+        serial_key = self.healtbox_manager.get_serial(ip)
+        if serial_key is not None:
+            try:
+                self.discovered_devices[serial_key] = HealthBox3Driver(ip=ip)
+                self.logger('Found Endura Delta device @ ip: {} with serial key: {}'.format(ip, serial_key))
+                self.register_ventilation_config(serial_key)
+            except Exception as ex:
+                self.logger("Discovered device @ {}, but could not connect to the device... {}".format(ip, ex))
+
+    def register_ventilation_config(self, serial_key):
+        # type: (str) -> None
+        """ Registers a new device to the gateway """
+        if serial_key not in self.discovered_devices:
+            self.logger('Could not register new ventilation device, serial key is not known to the plugin')
+            return
+        hbd = self.discovered_devices[serial_key]
+        if hbd is None:
+            self.logger('Could not register new ventilation device, driver is not working properly to request data')
+            return
+        serial_key = hbd.get_variable('serial')
+        config = {
+            "external_id": serial_key,
+            "source": {"type": "plugin", "name": HealthboxPlugin.name},
+            "name": hbd.get_variable('device name'),
+            "amount_of_levels": 4,
```

Comment:
Is this correct? Does the healthbox has 4 levels to set the ventilation box to? I know it is in the Endura Delta, but in the Healthbox I only knew of the "normal"  and the "boost" mode. It can be that the new firmware supports more levels, but not sure here.